File: frontend/src/api/equipment.ts
import request from './request';

export interface EquipmentData {
  id?: number;
  category: string;
  equipment_type: string;
  asset_code?: string;
  equipment_name: string;
  cabinet_model?: string;
  factory_number?: string;
  line_name?: string;
  station_name?: string;
  operation_date?: string;
  manufacturer?: string;
  province?: string;
  city?: string;
  district?: string;
  street?: string;
  address_detail?: string;
  longitude?: string;
  latitude?: string;
  customer_name?: string;
  remark?: string;
  extra_fields?: Record<string, any>;
}

export interface FieldOption {
  label: string;
  value: string;
  active: boolean;
}

export interface CascadeRule {
  field: string;
  mapping: Record<string, string[]>;
}

export interface VisibilityRule {
  field: string;
  value: string;
  action: 'show' | 'hide';
  operator?: string;
}

export interface RequiredRule {
  field: string;
  value: string;
  operator?: string;
}

export interface FieldConfig {
  id: number;
  field_name: string;
  field_label: string;
  field_type: string;
  is_required: string;
  options?: FieldOption[];
  max_length?: number;
  regex_pattern?: string;
  regex_hint?: string;
  min_value?: number;
  max_value?: number;
  decimal_places?: number;
  date_format?: string;
  default_value?: any;
  sort_order: number;
  is_active: string;
  parent_field_id?: number;
  visibility_rules?: VisibilityRule[];
  required_rules?: RequiredRule[];
  cascade_rules?: CascadeRule[];
}

export const equipmentApi = {
  list: (params?: any) => request.get('/api/equipment', { params }),
  get: (id: number) => request.get(`/api/equipment/${id}`),
  create: (data: EquipmentData) => request.post('/api/equipment', data),
  update: (id: number, data: EquipmentData) => request.put(`/api/equipment/${id}`, data),
  delete: (id: number) => request.delete(`/api/equipment/${id}`),
  uploadFile: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return request.post('/api/equipment/upload-photo', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  },

  // 字段配置管理
  getFieldConfigs: (includeDisabled?: boolean) =>
    request.get('/api/equipment/field-configs', { params: { include_disabled: includeDisabled || false } }),
  createFieldConfig: (data: any) =>
    request.post('/api/equipment/field-configs', data),
  updateFieldConfig: (id: number, data: any) =>
    request.put(`/api/equipment/field-configs/${id}`, data),
  deleteFieldConfig: (id: number) =>
    request.delete(`/api/equipment/field-configs/${id}`),
  manageOptions: (configId: number, data: { action: string; option?: FieldOption; old_value?: string }) =>
    request.post(`/api/equipment/field-configs/${configId}/options`, data),

  // 导入导出
  export: () => request.get('/api/equipment/export', { responseType: 'blob' }),
  import: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return request.post('/api/equipment/import', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  },
};

File: backend/schemas/equipment.py
from pydantic import BaseModel
from typing import Optional, Any
from datetime import datetime


class FieldConfigCreate(BaseModel):
    field_name: str
    field_label: str
    field_type: str
    is_required: str = "optional"
    options: Optional[list] = None
    max_length: Optional[int] = None
    regex_pattern: Optional[str] = None
    regex_hint: Optional[str] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    decimal_places: Optional[int] = None
    date_format: Optional[str] = None
    default_value: Optional[Any] = None
    sort_order: int = 0
    parent_field_id: Optional[int] = None
    visibility_rules: Optional[list] = None
    required_rules: Optional[list] = None
    cascade_rules: Optional[list] = None


class FieldConfigUpdate(BaseModel):
    field_name: Optional[str] = None
    field_label: Optional[str] = None
    field_type: Optional[str] = None
    is_required: Optional[str] = None
    options: Optional[list] = None
    max_length: Optional[int] = None
    regex_pattern: Optional[str] = None
    regex_hint: Optional[str] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    decimal_places: Optional[int] = None
    date_format: Optional[str] = None
    default_value: Optional[Any] = None
    sort_order: Optional[int] = None
    is_active: Optional[str] = None
    parent_field_id: Optional[int] = None
    visibility_rules: Optional[list] = None
    required_rules: Optional[list] = None
    cascade_rules: Optional[list] = None


class FieldConfigResponse(BaseModel):
    id: int
    field_name: str
    field_label: str
    field_type: str
    is_required: str
    options: Optional[list] = None
    max_length: Optional[int] = None
    regex_pattern: Optional[str] = None
    regex_hint: Optional[str] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    decimal_places: Optional[int] = None
    date_format: Optional[str] = None
    default_value: Optional[Any] = None
    sort_order: int
    is_active: str
    parent_field_id: Optional[int] = None
    visibility_rules: Optional[list] = None
    required_rules: Optional[list] = None
    cascade_rules: Optional[list] = None

    class Config:
        from_attributes = True


class OptionItem(BaseModel):
    """选项管理"""
    label: str
    value: str
    active: bool = True


class OptionManageRequest(BaseModel):
    """选项增删改请求"""
    action: str  # add, update, remove, toggle
    option: Optional[OptionItem] = None
    old_value: Optional[str] = None  # for update/remove, the value to find


class EquipmentCreate(BaseModel):
    category: str
    equipment_type: str
    asset_code: Optional[str] = None
    equipment_name: str
    cabinet_model: Optional[str] = None
    factory_number: Optional[str] = None
    line_name: Optional[str] = None
    station_name: Optional[str] = None
    operation_date: Optional[str] = None
    manufacturer: Optional[str] = None
    province: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None
    street: Optional[str] = None
    address_detail: Optional[str] = None
    longitude: Optional[str] = None
    latitude: Optional[str] = None
    customer_name: Optional[str] = None
    remark: Optional[str] = None
    extra_fields: Optional[dict] = None


class EquipmentUpdate(EquipmentCreate):
    pass


class EquipmentPhotoResponse(BaseModel):
    id: int
    photo_type: str
    file_path: str
    description: Optional[str] = None
    is_current: str
    created_at: datetime

    class Config:
        from_attributes = True


class AuditLogResponse(BaseModel):
    id: int
    field_name: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    modified_at: datetime

    class Config:
        from_attributes = True


class EquipmentResponse(BaseModel):
    id: int
    category: str
    equipment_type: str
    asset_code: Optional[str] = None
    equipment_name: str
    cabinet_model: Optional[str] = None
    factory_number: Optional[str] = None
    line_name: Optional[str] = None
    station_name: Optional[str] = None
    operation_date: Optional[str] = None
    manufacturer: Optional[str] = None
    province: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None
    street: Optional[str] = None
    address_detail: Optional[str] = None
    longitude: Optional[str] = None
    latitude: Optional[str] = None
    customer_name: Optional[str] = None
    remark: Optional[str] = None
    extra_fields: Optional[dict] = None
    photo_count: int = 0
    created_at: datetime
    updated_at: datetime
    photos: list[EquipmentPhotoResponse] = []
    audit_logs: list[AuditLogResponse] = []

    class Config:
        from_attributes = True


class EquipmentListResponse(BaseModel):
    total: int
    items: list[EquipmentResponse]
